From eb3257122183ac891f377d6b73343ae9ed5deab8 Mon Sep 17 00:00:00 2001
From: Peter Marschall <peter@adpm.de>
Date: Sun, 13 Mar 2011 21:41:12 +0100
Subject: [PATCH 01/32] OpenPGP: fix top-level DOs according to spec

Added:
* 00c4: new top-level DO in 2.0
        can also be found inside constructed DOs 006E/0073 in 2.0 & 1.1
* 0101: new optional top-level DO starting in 1.1
        for private use
        max 254 bytes;
        access: read - always; write - verify CHV2
* 0102: new optional top-level DO starting in 1.1
        for private use
        max 254 bytes;
        access: read - always; write - verify CHV3
* 5f52: new top-level DO in 2.0
        can also be found inside constructed DOs 006E in 2.0
* 7f21: new optional top-level DO in 2.0
        use: card holder certificate (e.g. X.509) for the AUT key in the card
Removed:
* 0073: never a top-level DO, but part of top-level constructed DO 006E
Changed:
* 005e: not a constructed DO, but a simple/primitive DO

Note:
Trying to read non-existent top-level DOs or top-level DOs that weren't defined
in a spec version later than the current card's version does not hurt.
They are returned as empty.
---
 src/libopensc/card-openpgp.c |    8 ++++++--
 1 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/libopensc/card-openpgp.c b/src/libopensc/card-openpgp.c
index 386e687..02a9561 100644
--- a/src/libopensc/card-openpgp.c
+++ b/src/libopensc/card-openpgp.c
@@ -91,12 +91,16 @@ static int		pgp_get_pubkey_pem(sc_card_t *, unsigned int,
 
 static struct do_info		pgp_objects[] = {
       {	0x004f,		0, 0,	sc_get_data,	sc_put_data	},
-      {	0x005e,		1, 0,	sc_get_data,	sc_put_data	},
+      {	0x005e,		0, 0,	sc_get_data,	sc_put_data	},
       {	0x0065,		1, 0,	sc_get_data,	sc_put_data	},
       {	0x006e,		1, 0,	sc_get_data,	sc_put_data	},
-      {	0x0073,		1, 0,	sc_get_data,	sc_put_data	},
       {	0x007a,		1, 0,	sc_get_data,	sc_put_data	},
+      {	0x00c4,		0, 0,	sc_get_data,	sc_put_data	},
+      {	0x0101,		0, 0,	sc_get_data,	sc_put_data	},
+      {	0x0102,		0, 0,	sc_get_data,	sc_put_data	},
       {	0x5f50,		0, 0,	sc_get_data,	sc_put_data	},
+      {	0x5f52,		0, 0,	sc_get_data,	sc_put_data	},
+      {	0x7f21,		1, 0,	sc_get_data,	sc_put_data	},
       { 0xb600,		1, 0,	pgp_get_pubkey,	NULL		},
       { 0xb800,		1, 0,	pgp_get_pubkey,	NULL		},
       { 0xa400,		1, 0,	pgp_get_pubkey,	NULL		},
-- 
1.7.4.4

